src/Symfony/Component/Security/Http/Impersonate/ImpersonateUrlGenerator.php

```diff
@@ -0,0 +1,129 @@
+    {
+        $this->requestStack = $requestStack;
+        $this->router = $router;
+        $this->tokenStorage = $tokenStorage;
+        $this->firewallMap = $firewallMap;
+    }
+
+    /**
+     * Generates the absolute logout path for the firewall.
+     **.
+     *
+     * @return string The Impersonate exit path
+     */
+    public function getImpersonateExitPath()
+    {
+        return $this->generateImpersonateExitUrl(UrlGeneratorInterface::ABSOLUTE_PATH);
+    }
+
+    /**
+     * Generates the absolute logout URL for the firewall.
+     *
+     * @return string The Impersonate exit URL
+     */
+    public function getImpersonateExitUrl()
+    {
+        return $this->generateImpersonateExitUrl(UrlGeneratorInterface::ABSOLUTE_URL);
+    }
+
+    /**
+     * Generates the Impersonate exit URL for the firewall.
+     *
+     * @param int $referenceType The type of reference (one of the constants in UrlGeneratorInterface)
+     *
+     * @return string The Impersonate exit URL
+     */
+    private function generateImpersonateExitUrl($referenceType)
+    {
+        $request = $this->requestStack->getCurrentRequest();
+        $exitPath = null;
+        if ($this->firewallMap instanceof FirewallMap) {
+            $firewallConfig = $this->firewallMap->getFirewallConfig($request);
+
+            // generate exit impersonation path from current request
+            if ($this->getImpersonatedUser() && null !== $switchUserConfig = $firewallConfig->getSwitchUser()) {
+                $exitPath = $request->getRequestUri();
+                $exitPath .= null === $request->getQueryString() ? '?' : '&';
+                $exitPath .= sprintf(
+                    '%s=%s',
+                    urlencode($switchUserConfig['parameter']),
+                    SwitchUserListener::EXIT_VALUE
+                );
+            }
+        }
+        if (null === $exitPath) {
+            throw new \LogicException('Unable to generate the impersonate exit URL without a path.');
+        }
+
+        if ('/' === $exitPath[0]) {
+            if (!$this->requestStack) {
+                throw new \LogicException('Unable to generate the impersonate exit URL without a RequestStack.');
+            }
+
+            $url = UrlGeneratorInterface::ABSOLUTE_URL === $referenceType ? $request->getUriForPath($exitPath) : $request->getBaseUrl().$exitPath;
+
+            if (!empty($parameters)) {
+                $url .= '?'.http_build_query($parameters);
+            }
+        } else {
+            if (!$this->router) {
+                throw new \LogicException('Unable to generate the impersonate exit URL without a Router.');
+            }
+
+            $url = $this->router->generate($exitPath, array(), $referenceType);
+        }
+
+        return $url;
+    }
+
+    private function getImpersonatedUser()
```

Comment:
I think this function is useful to know if the current user is being impersonated, so it should return a boolean value only and should be named something like `isImpersonatedUser()`?